rfcs/0051-limit-zkapp-cmds-per-block.md

```diff
@@ -0,0 +1,174 @@
+develop it, while it still wouldn't properly solve the problem.
+Therefore we should strive for a quick solution which already improves
+the situation and wait for the proper fix to come.
+
+## Prior art
+
+The problem of blockchain networks being unable to process incoming
+transactions fast enough is a well-known one and there are several
+techniques of dealing with it.
+
+One solution is to limit the block size (and hence indirectly the
+number of transactions fitting in a single block). The most notable
+example here is Bitcoin, which has a hard block size limit of 1MB.
+This is often criticized for limiting the network's throughput
+severely, but the restriction remains in place nonetheless, because
+the consequences of lifting it would be even worse.
+
+Mina also has its own block size limit, however, the problem we are
+dealing with here is different in that we've got two distinct
+categories of commands, only one of which is affected. Unfortunately,
+unless we move zkApp commands to a separate mempool, any limit set on
+zkApp commands throughput will also affect user commands by occupying
+mempool space (see Drawbacks above).
+
+Another solution is more related to execution time, especially that of
+smart contracts, which can - in principle - run indefinitely without
+termination and there is no easy way of preventing this without
+hindering expressiveness of a smart contract language significantly
+(due to insolvability of the halting problem). Major blockchains like
+Ethereum or Tezos, instead of limiting block size directly, restrict
+the number of computational steps (defined by some VM model) necessary
+to replay a block. A block which cannot be replayed in the specified
+number of steps is automatically considered invalid.
+
+The operation's execution time is modelled with gas. Each atomic
+computation is assigned a gas cost roughly proportional to the time
+the VM takes to execute that computation. Simultaneously, a block
+is given a hard gas limit and the total gas required by all the
+transactions within the block must be below that limit.
+
+Translating this solution to the discussed problem would involve
+modelling memory occupied by each operation in the scan state with
+some measure (analogous to gas) and then limiting the maximum value
+of operations (expressed in that measure) fitting in a block. This
+is a more complex solution than the one proposed here and probably
+requires significant time to devise the right model. It wouldn't
+also remove the problem of zkApp commands stacking in the mempool,
+although it might make it less severe by setting a more fine-grained
+limit. However, considering that it would still be a temporary
+solution, it's probably not worth the effort.
+
+## Unresolved questions
+
+Are the drawbacks described in this document an acceptable trade-off
+for preventing crashes due to out-of-memory issues? Is the
+alternative, more fine-grained solution viable?
+
+## Testing
```

Comment:
Can you add some unit tests in Staged ledger module to check that the limit is applied as expected with dependent payments and independent payments?

Reply:
Added.